### DISCUSSION TITLE
Invalid data type on Insert

### DISCUSSION THREAD

**Discussion by ANTVirGEO** [jackc/pgx#2101] [Q&A]
Good day! 

Previously I used default [database/sql](https://pkg.go.dev/database/sql) package to connect to my Postgres. 
And my Insert worked fine.
Now I changed the library to PGX (**jackc/pgx/v5 v5.6.0**) for the COPY method and got an interesting behavior: the code on my local system with Postgres 14.12 worked as it should and inserted all my data fine.
But the production version of Postgres 13 does something strange, giving me errors about bad insert types, as if either the columns or the data have swapped, and now the data is in the wrong columns.
`ERROR: invalid input syntax for type bigint: "[{"tipPs":"3","nomPs":"13183439","dlina":0}]" (SQLSTATE 22P02)
ERROR: invalid input syntax for type bigint: "2024-07-31 10:36:00.000000" (SQLSTATE 22P02)
ERROR: value too long for type character varying(3) (SQLSTATE 22001)
failed to encode args[0]: unable to encode 1722433220 into text format for text (OID 25): cannot find encode plan
ERROR: value "010002031070605" is out of range for type smallint (SQLSTATE 22003)
ERROR: value "125280" is out of range for type smallint (SQLSTATE 22003)`

In my DSN I previosly used _binary_parametres=yes_ 
`host=%s port=%d user=%s password=%s dbname=%s sslmode=%s connect_timeout=%s fallback_application_name=%s binary_parameters=yes`

And mine query uses SQUIRREL lib to make queries like this: 
//name of columns changes to them type in DB for this example (and Goland data type is same as in table)
```
query := sq.
  Insert("table").
  Columns(
	  "text",
	  "integer",
	  "smallint",
	  "varchar(3)",
	  "bool",
	  "timestamp",
  )

for _, d := range data.Data {
  query = query.Values(
	  d.SomeText,
	  d.SomeInteger,
	  d.SomeSmallInt,
	  d.3CharText,
	  d.SomeBool,
	  d.SomeTimestamp
  )
}

querySql, args, err := query.PlaceholderFormat(sq.Dollar).ToSql()
if err != nil {
  return 0, err
}

result, err := pool.Exec(context.Background(), querySql, args...)
if err != nil {
  return 0, err
}

return result.RowsAffected(), nil
```

And with database/sql was all the same, but end was:
```
result, err := db.ExecContext(context.Background(), querySql, args...)
if err != nil {
  return 0, err
}

return result.RowsAffected()
```

And mine connection here:
```
func GetPoolConnection (some data) {
  dsn := fmt.Sprintf(
  "host=%s port=%d user=%s password=%s dbname=%s sslmode=%s connect_timeout=%s "+
	  "default_query_exec_mode=simple_protocol",
  p.Host,
  p.Port,
  p.Username,
  p.Password,
  p.DBName,
  sslModeStr,
  connectTimoutStr)

  config, err := pgxpool.ParseConfig(dsn)
  if err != nil {
    return nil, fmt.Errorf("unable to parse DSN: %w", err)
  }
  
  config.MaxConns = 3
  config.MinConns = 1
  config.MaxConnLifetime = time.Minute * 5
  config.MaxConnIdleTime = time.Minute
  
  pool, err := pgxpool.NewWithConfig(ctx, config)
  if err != nil {
    return nil, fmt.Errorf("unable to create connection pool: %w", err)
  }

  return pool
}
```

I've searched around, but I can't find a reason why this is happening.
Maybe there is an additional setting needed somewhere in the connection, or the data needs to be processed differently in the call itself? It's confusing that all this works locally.
Please tell me, maybe you have some ideas on what to look for in my case?
Thanks!

**Comment by jackc** [comment]
I'm not sure what squirrel is doing. But I would suggest calling `CopyFrom` directly and seeing if that works.
